File: app/src/main/java/me/liuyun/bjutlgn/ui/ThemeActivity.java
package me.liuyun.bjutlgn.ui;

import android.app.Activity;
import android.content.res.Resources;
import android.graphics.drawable.GradientDrawable;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.RadioButton;
import android.widget.TextView;

import butterknife.BindView;
import butterknife.ButterKnife;
import me.liuyun.bjutlgn.R;
import me.liuyun.bjutlgn.util.ThemeRes;
import me.liuyun.bjutlgn.util.ThemeHelper;

public class ThemeActivity extends AppCompatActivity {
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_theme);

        Toolbar toolbar = ButterKnife.findById(this, R.id.theme_toolbar);
        setSupportActionBar(toolbar);
        toolbar.setNavigationOnClickListener(view -> onBackPressed());

        RecyclerView recyclerView = ButterKnife.findById(this, R.id.recycler);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setAdapter(new ThemeAdapter());
    }

    class ThemeAdapter extends RecyclerView.Adapter<ThemeAdapter.ThemeItemHolder> {
        Resources resources = getResources();
        Resources.Theme theme = getTheme();
        int white = resources.getColor(android.R.color.white, theme);
        int grey = resources.getColor(android.R.color.darker_gray, theme);

        @Override
        public ThemeItemHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.theme_item, parent, false);
            return new ThemeItemHolder(view);
        }

        @Override
        public void onBindViewHolder(ThemeItemHolder holder, int position) {
            ThemeRes res = ThemeRes.values()[position];
            int themeColor = resources.getColor(res.getColor(), theme);
            holder.themeId = res.getStyle();
            holder.tvTheme.setText(res.getName());
            holder.tvTheme.setTextColor(themeColor);
            if (ThemeHelper.getInstance().getCurrentStyle() == res.getStyle()) {
                holder.btnChoose.setChecked(true);
                holder.btnChoose.setTextColor(themeColor);
                holder.btnChoose.setText(R.string.theme_using);
                holder.colorDot.setTextColor(white);
            } else {
                holder.btnChoose.setChecked(false);
                holder.btnChoose.setTextColor(grey);
                holder.btnChoose.setText(R.string.theme_use);
                holder.colorDot.setTextColor(themeColor);
            }
            GradientDrawable gradientDrawable = new GradientDrawable();
            gradientDrawable.setColor(themeColor);
            gradientDrawable.setCornerRadius(40f);
            holder.colorDot.setBackground(gradientDrawable);
        }

        @Override
        public int getItemCount() {
            return ThemeRes.values().length;
        }

        class ThemeItemHolder extends RecyclerView.ViewHolder {
            @BindView(R.id.color_dot) TextView colorDot;
            @BindView(R.id.tv_title) TextView tvTheme;
            @BindView(R.id.btn_choose) RadioButton btnChoose;
            private int themeId;

            ThemeItemHolder(View itemView) {
                super(itemView);
                ButterKnife.bind(this, itemView);
                btnChoose.setOnClickListener(view -> onItemClick());
            }

            private void onItemClick() {
                ThemeHelper.getInstance().setTheme((Activity) itemView.getContext(), themeId);
            }
        }
    }
}

File: app/src/main/kotlin/me/liuyun/bjutlgn/ui/ThemeActivity.kt
package me.liuyun.bjutlgn.ui

import android.graphics.drawable.GradientDrawable
import android.os.Bundle
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.RecyclerView
import me.liuyun.bjutlgn.R
import me.liuyun.bjutlgn.databinding.ActivityThemeBinding
import me.liuyun.bjutlgn.databinding.ItemThemeBinding
import me.liuyun.bjutlgn.util.ThemeHelper
import me.liuyun.bjutlgn.util.ThemeRes

class ThemeActivity : AppCompatActivity() {
    private lateinit var binding: ActivityThemeBinding

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityThemeBinding.inflate(layoutInflater)
        setContentView(binding.root)
        setSupportActionBar(binding.appBar.toolbar)
        binding.appBar.toolbar.setNavigationOnClickListener { onBackPressed() }

        binding.recycler.adapter = ThemeAdapter()
    }

    internal inner class ThemeAdapter : RecyclerView.Adapter<ThemeAdapter.ThemeItemHolder>() {
        var white = resources.getColor(android.R.color.white, theme)
        var grey = resources.getColor(android.R.color.darker_gray, theme)

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
                ThemeItemHolder(ItemThemeBinding.inflate(LayoutInflater.from(parent.context), parent, false))

        override fun onBindViewHolder(holder: ThemeItemHolder, position: Int) {
            val res = ThemeRes.values()[position]
            val themeColor = resources.getColor(res.color, theme)
            holder.themeId = res.style
            holder.binding.tvTitle.setText(res.themeName)
            holder.binding.tvTitle.setTextColor(themeColor)
            holder.binding.btnChoose.setOnClickListener { ThemeHelper.setTheme(this@ThemeActivity, holder.themeId) }
            if (ThemeHelper.currentStyle == res.style) {
                holder.binding.btnChoose.isChecked = true
                holder.binding.btnChoose.setTextColor(themeColor)
                holder.binding.btnChoose.setText(R.string.theme_using)
                holder.binding.colorDot.setTextColor(white)
            } else {
                holder.binding.btnChoose.isChecked = false
                holder.binding.btnChoose.setTextColor(grey)
                holder.binding.btnChoose.setText(R.string.theme_use)
                holder.binding.colorDot.setTextColor(themeColor)
            }
            val gradientDrawable = GradientDrawable()
            gradientDrawable.setColor(themeColor)
            gradientDrawable.cornerRadius = 40f
            holder.binding.colorDot.background = gradientDrawable
        }

        override fun getItemCount() = ThemeRes.values().size

        internal inner class ThemeItemHolder(val binding: ItemThemeBinding, var themeId: Int = 0) : RecyclerView.ViewHolder(binding.root)
    }
}
